src/transforms/transaction.rs

```diff
@@ -0,0 +1,812 @@
+                        Some((k, v.clone().into()))
+                    }
+                })
+                .collect(),
+        }
+    }
+
+    fn add_event(&mut self, e: LogEvent, strategies: &IndexMap<String, MergeStrategy>) {
+        for (k, v) in e.all_fields() {
+            let strategy = strategies.get(&k);
+            match self.fields.entry(k) {
+                hash_map::Entry::Vacant(entry) => {
+                    if let Some(strat) = strategy {
+                        match get_value_merger(v.clone(), strat) {
+                            Ok(m) => {
+                                entry.insert(m);
+                            }
+                            Err(err) => {
+                                warn!("failed to merge value: {}", err);
+                            }
+                        }
+                    } else {
+                        entry.insert(v.clone().into());
+                    }
+                }
+                hash_map::Entry::Occupied(mut entry) => {
+                    if let Err(err) = entry.get_mut().add(v.clone()) {
+                        warn!("failed to merge value: {}", err);
+                    }
+                }
+            }
+        }
+        self.stale_since = Instant::now();
+    }
+
+    fn flush(mut self) -> LogEvent {
+        let mut event = Event::new_empty_log().into_log();
+        for (k, v) in self.fields.drain() {
+            if let Err(err) = v.insert_into(k, &mut event) {
+                warn!("failed to merge values for field: {}", err);
+            }
+        }
+        event
+    }
+}
+
+//------------------------------------------------------------------------------
+
+pub struct Transaction {
+    expire_after: Duration,
+    flush_period: Duration,
+    identifier_fields: Vec<Atom>,
+    merge_strategies: IndexMap<String, MergeStrategy>,
+    transaction_merge_states: HashMap<Discriminant, TransactionState>,
+    ends_when: Option<Box<dyn Condition>>,
+}
+
+impl Transaction {
+    fn new(config: &TransactionConfig) -> crate::Result<Self> {
+        let ends_when = if let Some(ends_conf) = &config.ends_when {
+            Some(ends_conf.build()?)
+        } else {
+            None
+        };
+
+        let identifier_fields = config
+            .identifier_fields
+            .clone()
+            .into_iter()
+            .map(Atom::from)
+            .collect();
+
+        Ok(Transaction {
+            expire_after: Duration::from_millis(config.expire_after_ms.unwrap_or(30000)),
+            flush_period: Duration::from_millis(config.flush_period_ms.unwrap_or(1000)),
+            identifier_fields,
+            merge_strategies: config.merge_strategies.clone(),
+            transaction_merge_states: HashMap::new(),
+            ends_when,
+        })
+    }
+
+    fn flush_into(&mut self, output: &mut Vec<Event>) {
+        let mut flush_discriminants = Vec::new();
+        for (k, t) in &self.transaction_merge_states {
```

Comment:
I wish we could find a way to iter over the merge states by expiration time and stop when we know all future states are not expired. :|

This has potentially harmful performance if it ends up looping over large amounts of discriminants.

Reply:
Yes I think there's an alternate implementation based with on a `DelayQueue` or our `ExpiringHashMap`. I think that's something we can tackle in a followup PR.